### ISSUE TITLE
[no-invalid-void-type] jest.fn<void, []> should be allowed

### ISSUE THREAD

**Issue by yoursunny** [typescript-eslint/typescript-eslint#2044]
**Repro**

test.ts
```ts
test("hello", async () => {
  const fn = jest.fn<void, [number]>();
  fn(1);
  expect(fn).toHaveBeenCalledWith(1);
});
```

.eslintrc.js
```js
module.exports = {
  root: true,
  parser: '@typescript-eslint/parser',
  plugins: [
    '@typescript-eslint',
  ],
  rules: {
    '@typescript-eslint/no-invalid-void-type': 'error',
  },
};
```

package.json
```json
{
  "scripts": {
    "build": "tsc --noEmit",
    "lint": "eslint test.ts",
    "test": "jest"
  },
  "devDependencies": {
    "@types/jest": "25.2.3",
    "@typescript-eslint/eslint-plugin": "^2.34.1-alpha.0",
    "@typescript-eslint/parser": "2.34.0",
    "eslint": "7.0.0",
    "jest": "26.0.1",
    "ts-jest": "26.0.0",
    "typescript": "3.9.2"
  }
}
```

tsconfig.json
```jsonc
{
  "compilerOptions": {
    "module": "CommonJS",
    "moduleResolution": "Node",
    "strict": true,
    "target": "ESNext",
    "types": [
      "jest"
    ]
  }
}
```

jest.config.js
```js
module.exports = {
  preset: 'ts-jest',
  testEnvironment: 'node',
};
```

**Expected Result**

ESLint should report no error.

**Actual Result**

ESLint reports error:
```
C:\no-invalid-void-type\test.ts
  2:22  error  void is only valid as a return type or generic type variable  @typescript-eslint/no-invalid-void-type      

✖ 1 problem (1 error, 0 warnings)
```

**Additional Info**

[jest.fn()](https://jestjs.io/docs/en/jest-object#jestfnimplementation) creates a mock function.
Its typing is [declared](https://unpkg.com/browse/@types/jest@25.2.3/index.d.ts#L159) as:
```ts
function fn<T, Y extends any[]>(implementation?: (...args: Y) => T): Mock<T, Y>;
```
in which:
* `T` is the return type of the mock function.
* `Y` is a tuple of parameter types.

In my test.ts, `jest.fn<void, []>` creates a mock function that accepts no parameter and returns `void`. This should be a valid usage of `void` type because it represents a return type.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.34.0` and canary |
| `@typescript-eslint/parser`        | `2.34.0` |
| `TypeScript`                       | `3.9.2` |
| `ESLint`                           | `7.0.0` |
| `node`                             | `14.2.0` |
| `npm`                              | `6.14.0` |

**Comment by bradzacher** [comment]
> Disallows usage of void type outside of return types or a generic type's type parameters. If void is used as return type, it shouldn’t be a part of intersection/union type with most other types.

You are using `void` as a type parameter in a function call, which is explicitly banned by this rule.

Test code often intentionally works around lint rules because testing is messy business.
My suggestion is to either use a disable comment for this, or disable the rule for test files.
